### PR TITLE
enhancement(http sink): Retrying the HTTP sink in case of 404s and request timeouts

### PR DESCRIPTION
**Pull request by noble-varghese**
Adding the functionality to retry the log sending on getting a request timeout (408) from the HTTP sink. Since the default behavior is to drop the events during such events, this would affect the log propagation to the sinks.

solves: #10870

### CODE REVIEW

**Comment by pront** [comment]
Hi @noble-varghese, thank you for the contribution. I think retrying for these codes is reasonable, I will take a closer look tomorrow. 

Note: Need to update the docs as well. See https://vector.dev/docs/reference/configuration/sinks/http/#retry-policy.

**Comment by noble-varghese** [comment]
@pront Thanks for pointing it out ! I have updated the docs page and added unit tests on the http utils as well :)

**Review by brett0000FF — APPROVED** [review]

**Review by pront — CHANGES_REQUESTED** [review]
Hey @noble-varghese, I noticed the E2E test `http_to_http` [failed](https://github.com/vectordotdev/vector/actions/runs/11257844072/job/31314324260?pr=21457). This change actually is a breaking one. Since it is hard to evaluate the risk, it is better to err on the side of caution. After some discussion with the team, I created https://github.com/vectordotdev/vector/issues/21469 to layout a solution for this.

**Comment by pront** [comment]
To unblock this PR, perhaps we can add 408 to the defaults? Still a breaking change, but seems desirable and low risk. cc @jszwedko

File: src/sinks/util/http.rs [hunk 1]
```diff
@@ -520,6 +523,9 @@ where
 
         match status {
             StatusCode::TOO_MANY_REQUESTS => RetryAction::Retry("too many requests".into()),
+            StatusCode::NOT_FOUND => RetryAction::Retry("not found".into()),
+            StatusCode::REQUEST_TIMEOUT => RetryAction::Retry("request timeout".into()),
+            StatusCode::FORBIDDEN => RetryAction::Retry("request forbidden".into()),
```
Comment (pront) [inline src/sinks/util/http.rs:528]:
```suggestion
            StatusCode::REQUEST_TIMEOUT => RetryAction::Retry("request timeout".into()),
```

**Comment by pront** [comment]
I tried to push directly to the fork but I don't have write permissions.

**Comment by noble-varghese** [comment]
> I tried to push directly to the fork but I don't have write permissions.

I will push those changes. Thanks !

**Comment by noble-varghese** [comment]
@pront  Just curious to know, how do you run these tests on the local. It was taking me about 30-40mins for it to run. Is there a faster way ?

File: changelog.d/21457_more_error_codes_on_http_sink.enhancement.md [hunk 2]
```diff
@@ -0,0 +1,3 @@
+The `http sink` will now retry requests when the status is 408(request timeout).
```
Comment (jszwedko) [inline changelog.d/21457_more_error_codes_on_http_sink.enhancement.md:1]:
```suggestion
The `http sink` will now retry requests when the response is a request timeout (HTTP 408).
```

Small proofreading.

**Comment by jszwedko** [comment]
> @pront Just curious to know, how do you run these tests on the local. It was taking me about 30-40mins for it to run. Is there a faster way ?

You could try `cargo test --no-default-features --features sinks-http`. This and other tips are mentioned here: https://github.com/vectordotdev/vector/blob/master/docs/DEVELOPING.md#tips-and-tricks

**Comment by pront** [comment]
> > @pront Just curious to know, how do you run these tests on the local. It was taking me about 30-40mins for it to run. Is there a faster way ?
> 
> You could try `cargo test --no-default-features --features sinks-http`. This and other tips are mentioned here: https://github.com/vectordotdev/vector/blob/master/docs/DEVELOPING.md#tips-and-tricks

Thanks Jesse, very neat.

Also, I always run the following before pushing to the remote branch:

- cargo install -f --path vdev #cd to the repo root
- `cargo vdev fmt`
- `cargo vdev check rust --clippy`

**Review by pront — APPROVED** [review]

**Pull request merged**